toolkit/pallets/block-participation/src/lib.rs

```diff
@@ -0,0 +1,113 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+use frame_support::pallet_prelude::*;
+pub use pallet::*;
+use sp_block_participation::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_system::pallet_prelude::*;
+	use scale_info::prelude::format;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_block_production_log::Config {
+		/// Type identifying indirect block production participants on the Partner Chain
+		/// This can be native stakers on Partner Chain, stakers on the main chain or other.
+		type BlockShareholderId: Member + Parameter + MaxEncodedLen;
+
+		/// Should return slot up to which block production data should be released or None.
+		fn should_release_data(slot: Slot) -> Option<Slot>;
+
+		/// Inherent ID for which block participation data should be provided.
+		/// It should be set to the ID used by the pallet that will process participation data for
+		/// paying out block rewards or other purposes.
+		const TARGET_INHERENT_ID: InherentIdentifier;
+	}
+
+	#[pallet::inherent]
+	impl<T: Config> ProvideInherent for Pallet<T> {
+		type Call = Call<T>;
+		type Error = sp_block_participation::InherentError;
+		const INHERENT_IDENTIFIER: InherentIdentifier = sp_block_participation::INHERENT_IDENTIFIER;
+
+		fn create_inherent(data: &InherentData) -> Option<Self::Call> {
+			let up_to_slot = Self::decode_inherent_data(data)?;
+
+			Some(Call::note_processing { up_to_slot })
+		}
+
+		fn check_inherent(call: &Self::Call, data: &InherentData) -> Result<(), Self::Error> {
+			let Some(expected_inherent_data) = Self::decode_inherent_data(data) else {
+				return Err(Self::Error::UnexpectedInherent);
+			};
+
+			let Self::Call::note_processing { up_to_slot } = call else {
+				unreachable!("There should be no other extrinsic in the pallet")
+			};
+
+			if *up_to_slot != expected_inherent_data {
+				return Err(Self::Error::IncorrectSlotBoundary);
+			}
+
+			Ok(())
+		}
+
+		fn is_inherent(call: &Self::Call) -> bool {
+			matches!(call, Call::note_processing { .. })
+		}
+
+		fn is_inherent_required(data: &InherentData) -> Result<Option<Self::Error>, Self::Error> {
+			if Self::decode_inherent_data(data).is_some() {
+				Ok(Some(Self::Error::InherentRequired))
+			} else {
+				Ok(None)
+			}
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		fn decode_inherent_data(data: &InherentData) -> Option<Slot> {
+			data.get_data(&Self::INHERENT_IDENTIFIER)
+				.map_err(|err| {
+					format!("Block participation inherent data is present but invalid: {err:?}")
+				})
+				.unwrap()
+		}
```

Comment:
here would be great to add new error type for this pallet with `#[pallet::error]`, it will allow us to change return type to `Result<Option<Slot>, Error<T>>` and remove unwrap
```suggestion
	impl<T: Config> Pallet<T> {
		fn decode_inherent_data(data: &InherentData) -> Result<Option<Slot>, Error<T>> {
			data.get_data(&Self::INHERENT_IDENTIFIER)
				.map_err(|err| Error::<T>::InvalidParticipationInherentData)
		}
	}
```

Reply:
added. that still requires unwrapping when proposing a block, but we can't do anything about it